### PR TITLE
Prometheus support

### PR DESCRIPTION
**Pull request by mtanda**
I wrote Prometheus plugin mentioned in https://github.com/grafana/grafana/issues/1390.
It implements these features.
- Query editor with method name auto completion
- Group by label of metrics
- Auto calibrate resolution of metrics

Please review this PR.

prometheus_edit_query
prometheus_graph

### CODE REVIEW

**Comment by syepes** [comment]
+1

**Comment by mtanda** [comment]
There is a problem fetching too many data points, it can't draw on graph.
Follow advice in this comment.
https://github.com/grafana/grafana/pull/1738#issuecomment-91597438
I fix it, use maxDataPoints to calibrate query step option.

**Comment by jfcervera** [comment]
Great, just what I was looking for! +1

**Comment by juliusv** [comment]
@mtanda Hey, Prometheus author here. Great to see Prometheus support in Grafana being added!

I hope to try this out soon, but some questions, since I don't know Grafana too well:

- in the first screenshot, there are two input fields: "Query" and "Metric". How do these go together? Usually you provide an entire query expression when getting data out of Prometheus, not just a metric name?
- what does the "Labels" input do if I enter an arbitrary Prometheus expression in the "Query" field, not just a metric name? Does it actually try to modify the query to Prometheus somehow still?
- is there a way already to format the legend strings differently (as in PromDash, where you can assign custom legend format strings to show only certain labels)? If not, is this feasible to add later?

File: datasources/prometheus/plugin.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  "pluginType": "datasource",
+  "name": "Prometheus",
+
+  "type": "prometheus",
+  "serviceName": "PrometheusDatasource",
+
+  "module": "plugins/datasource/prometheus/datasource",
+
+  "partials": {
+    "config": "app/plugins/datasource/prometheus/partials/config.html",
```
Comment (jfcervera) [inline datasources/prometheus/plugin.json]: Haven't looked at the project in detail and maybe part of the build process copies files from some places to others and what I'm saying won't apply, but one thing that strikes me is that the partial files are configured in the path app/plugins/datasource when the corresponding files below are created in app/plugins/datasources ... is that normal?

File: datasources/prometheus/plugin.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    "query": "app/plugins/datasource/prometheus/partials/query.editor.html"
```
Comment (jfcervera) [inline datasources/prometheus/plugin.json]: Same as above comment

**Comment by mtanda** [comment]
@juliusv Hi, thank you for having interest in this PR.
> in the first screenshot, there are two input fields: "Query" and "Metric". How do these go together? Usually you provide an entire query expression when getting data out of Prometheus, not just a metric name?

In fact, the "Metric" field is not input field.
It works like metric select box of Prometheus Expression browser, assist in inputting metric name to "Query" field.

> what does the "Labels" input do if I enter an arbitrary Prometheus expression in the "Query" field, not just a metric name? Does it actually try to modify the query to Prometheus somehow still?

At first, I want to implement such UI, but I noticed the Prometheus query syntax is little complicated, It is hard for me..
So, the "Labels" field is not used for constructing query. It affects only the name of each legend.

In the second screenshot, "handler" is specified in "Labels", the legend name looks like "http_request_duration_microseconds{handler=foobar}”.
If “handler” is not specified, the name looks like "http_request_duration_microseconds".
The “Labels” controls what labels included in legend name.
(The query expression might be 'http_request_duration_microseconds{handler=~".*"}')

> is there a way already to format the legend strings differently (as in PromDash, where you can assign custom legend format strings to show only certain labels)? If not, is this feasible to add later?

I'm not sure I understand the feature, but it might be provided by the "Labels" field.

I’m site reliability engineer, and trying to use Prometheus for current project.
I already evaluated Grafana before using Prometheus, I know the Grafana provides many features I want to use.
So I need to support Prometheus by Grafana.
Any feedback is welcome, I’m waiting for your feedback.

File: datasources/prometheus/plugin.json (hunk 1 above)
Comment (mtanda) [inline datasources/prometheus/plugin.json]: @jfcervera Sorry, I don't know the detail of Grafana build process.

I look thourgh Grunt tasks, I found the copy tasks.
https://github.com/grafana/grafana/blob/master/tasks/options/copy.js
This task copy the all Grafana source from original directory to temp directory.
Does that answer your question?

Please let me add more information.
To use this plugin, you need to copy this plugin to Grafana source tree and build Grafana by yourself.
I follow the step described in this Dockerfile.
https://github.com/grafana/grafana-docker/blob/e3d04fc7d509092fd7abe0291188e502b9f45924/Dockerfile

I usually execute following commands.
```
git clone https://github.com/grafana/grafana.git
git clone https://github.com/grafana/grafana-plugins.git
cp -a grafana-plugins/datasources/prometheus grafana/public/app/plugins/datasource
cd grafana
go run build.go setup
go run build.go build
npm install
npm install -g grunt-cli
grunt release
```

The "go run build.go build" may register this plugin to Golang Grafana backend, so you need to copy plugin before executing it.

**Comment by jfcervera** [comment]
Thanks @mtanda. The build process must do the moving as you say. Given the nature of the files I just copied them manually (into the datasource directory) on 2.0.0-beta3 and it worked fine without doing a build.

Thanks for creating the plug-in.

**Comment by torkelo** [comment]
Yes, you do not need to do a build. You need to restart the backend, it will scan for the plugin.json files during startup.

**Comment by mtanda** [comment]
Ah! I understand the loadPluginJson() register plugin code, no need to build it!
@jfcervera @torkelo Thank you for telling me.

**Comment by juliusv** [comment]
@mtanda Ah ok, thanks for clearing that up.

The way legend format strings work in PromDash is like the following. Let's say you get back the following time series from a Prometheus query `http_requests_total{job="api-server"}`:

```
http_requests_total{job="api-server",instance="1.2.3.4:9000",method="POST"}
http_requests_total{job="api-server",instance="1.2.3.4:9000",method="GET"}
http_requests_total{job="api-server",instance="5.6.7.8:9000",method="POST"}
http_requests_total{job="api-server",instance="5.6.7.8:9000",method="GET"}
```

The metric name and the `job` label value is the same for all of these time series, so there is little value in displaying them all in each series's legend name. So PromDash allows you provide a format string for every Prometheus query expression (and thus result). For example, the format string:

    {{instance}} - {{method}}

...would result in legend names that look like this:

    1.2.3.4:9000 - POST
    1.2.3.4:9000 - GET
    5.6.7.8:9000 - POST
    5.6.7.8:9000 - GET

You can mix literal string characters with `{{}}` interpolations in a legend format string and it will build the actual legend strings from it. This is quite useful as otherwise the complete time series names can get quite long (like in your example screenshot - although only one of the labels is included, the metric name is always repeated). Would be great to support a similar thing in Grafana.

**Comment by mtanda** [comment]
@juliusv Thanks for explanation.
It's better than current display style, I'll implement it.

**Comment by juliusv** [comment]
@mtanda Awesome! Now I just have to figure out how to use grafana with these plugins...

**Comment by mtanda** [comment]
@juliusv Sorry for my late reply...
There are pre-compiled Grafana binaries.
http://grafana.org/download/

Installing these binaries is fast and easy.

If using the binary tar, execute following commands.
```
$ curl -L -O https://grafanarel.s3.amazonaws.com/builds/grafana-2.0.1.linux-x64.tar.gz
$ tar zxf grafana-2.0.1.linux-x64.tar.gz
$ git clone https://github.com/grafana/grafana-plugins.git
$ cp -a grafana-plugins/datasources/prometheus grafana-2.0.1/public/app/plugins/datasource/
$ cd grafana-2.0.1/
$ ./bin/grafana-server web
```

If using the deb package, execute following commands.
```
$ curl -L -O https://grafanarel.s3.amazonaws.com/builds/grafana_2.0.1_amd64.deb
$ sudo dpkg -i grafana_2.0.1_amd64.deb
$ git clone https://github.com/grafana/grafana-plugins.git
$ sudo cp -a grafana-plugins/datasources/prometheus /usr/share/grafana/public/app/plugins/datasource/
$ sudo service grafana-server start
```

To add Prometheus data source, this document (for OpenTSDB) is good reference.
http://docs.grafana.org/datasources/opentsdb/

**Comment by juliusv** [comment]
@mtanda Thanks for the detailed instructions! I just tried it out now, great work!

Some initial feedback:
- Prometheus query expressions can get quite long. The current query input field is quite small. It'd be great to make it much longer so that you can see more context when typing a longer expression (for example, we have queries in PromDash like `count(sum(bazooka_instance_memory_usage_bytes) by (app, proc) / 1024 / 1024 / 1024 < 64)` and even much longer ones).
- When simply querying for `prometheus_local_storage_chunk_ops_total`, which has a `type` label (confirmed also by looking at the AJAX query response data in the Chrome inspector), specifying the `type` label in the "Labels" input doesn't add it to the legend (am I doing something wrong?):

grafana

- One feature I missed from PromDash is the ability to jump from the PromDash/Grafana expression input directly to the equivalent graph/tabular view on the Prometheus server itself. This gives you quick access especially to the tabular view, which makes it easier to construct queries iteratively, starting from a large unfiltered dataset (can be queried cheaper in tabular view) and then filter/aggregate it down to a graphable dataset. In PromDash, the button for that looks like this:

link_to_graph

- Currently every time I edit the expression (or simply focus/unfocus it without editing), Grafana sends the same query to Prometheus *twice* instead of just once, as I noticed when looking at the requests in the Chrome inspector.

**Comment by mtanda** [comment]
Thanks for feedback.

> * The current query input field is quite small.

I totally agree with you :-)
I'll expand input field.

> * "Labels" input doesn't add it to the legend

I apologize for any confusion.
After input field, need to click "add Label" button.
Anyway, this feature might be replaced by "legend format strings".

> * jump from the PromDash/Grafana expression input directly to the equivalent graph/tabular view on the Prometheus server itself.

Okay, I'll add the link.

> * Grafana sends the same query to Prometheus twice

I'm not sure why Grafana sends the same query twice...
I'll check the plugin code and Grafana code.

**Comment by juliusv** [comment]
>
>
>    - jump from the PromDash/Grafana expression input directly to the
>    equivalent graph/tabular view on the Prometheus server itself.
>
>  Okay, I'll add the link.
>

Great! To do that, you'll need to encode the right graph options and query
into the hash part of the Prometheus URL. This is how PromDash does it, if
you need inspiration:

https://github.com/prometheus/promdash/blob/5496e445f26119dab29fda47d71c49ea8a1fd30c/app/assets/javascripts/angular/directives/expression.js.erb#L44-L78

You probably care only about the last couple of lines of that (the initial
part of that function contain a lot of PromDash-specific code):

https://github.com/prometheus/promdash/blob/5496e445f26119dab29fda47d71c49ea8a1fd30c/app/assets/javascripts/angular/directives/expression.js.erb#L67-L78

**Comment by mtanda** [comment]
@juliusv I'll refer those links. Thank you!

**Pull request merged**